File: solute/epfl/widgets/slider/slider.py
# -*- coding: utf-8 -*-

from solute.epfl.core import epflwidgetbase

class SliderWidget(epflwidgetbase.WidgetBase):


    """ Displays a horizontal slider element.
    min_value and max_value can be integer or float
    on_change can be a event-handler-name or "submit" which then submittes the complete form

    Additional jinja-attributes:

    {{ field.value_box }}: If used in template it shows the current value of the slider

    """


    name = "slider"
    template_name = "slider/slider.html"
    asset_spec = "solute.epfl.widgets:slider/static"

    js_name = ["slider.js"]
    css_name = []

    param_def = {"min_value": epflwidgetbase.NumberType,
                 "max_value": epflwidgetbase.NumberType,
                 "on_change": epflwidgetbase.EventType}


    def update_data_source(self, data_source):

        if type(data_source.params["min_value"]) is float:
            data_source.value_divisor = 10000
            data_source.params["min_value"] *= data_source.value_divisor
            data_source.params["max_value"] *= data_source.value_divisor
        else:
            data_source.value_divisor = 1

    def pre_render(self):
        super(SliderWidget, self).pre_render()

        self.field.value_box = "<div id='" + self.field.name + "_value_box'></div>"

File: solute/epfl/widgets/slider/static/slider.js

epfl.init_slider_widget = function(params) {
	$(document).ready(function() {

		function show_value(value) {
			if (params["value_divisor"] == 1) {
				value = parseInt(value);
			} else {
				value = value.toFixed(2);
			}
			$("#" + params.name + "_value_box").html(value);
		};

		var value = parseFloat($("#" + params.name).val());
		show_value(value);
		var slider = $("#" + params.name + "_slider").slider({

			min: params.min,
			max: params.max,
			value: value * params["value_divisor"],
			
			slide: function(event, ui) {
				var value = ui.value / params["value_divisor"];
				show_value(value);
			},
			change: function(event, ui) {
				var value = ui.value / params["value_divisor"];
				$("#" + params.name).val(value);
				show_value(value);
				if (params.on_change == "submit") {
					$("#" + params.name).closest("form").submit();
				} else if (params.on_change) {
					var event = epfl.make_component_event(params.cid, params.on_change, {"widget_name": params.name,
																					     "value": value});
					epfl.send(event);
				};
			}
		});
	});
};
